### ISSUE TITLE
SIGSEGV in PhaseIdealLoop::build_loop_late_post_work

### ISSUE THREAD

**Issue by tivervac** [adoptium/adoptium-support#530]
## Summary

We run an Eclipse-based product obfuscated using [ZKM](https://www.zelix.com/klassmaster/index.html). Running our tests on CI has been causing frequent SIGSEGV's.

## Steps to reproduce

The error is "rare" (one in 10 builds usually, on some of our branches it's 3/4, on others 1/20),

## Expected results

No crash

## Actual results

Random SIGSEGV's likely heavily influenced by code layout and timings.

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007ff7086a270b, pid=27153, tid=27169
#
# JRE version: OpenJDK Runtime Environment Temurin-17.0.3+7 (17.0.3+7) (build 17.0.3+7)
# Java VM: OpenJDK 64-Bit Server VM Temurin-17.0.3+7 (17.0.3+7, mixed mode, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# V  [libjvm.so+0xac870b]  PhaseIdealLoop::build_loop_late_post_work(Node*, bool)+0x13b
#
# Core dump will be written. Default location: /home/jenkins/agent/workspace/line_8382-speed-up-product-build/com.sigasi.hdt.vhdl.test.projects/core.27153
#
# An error report file with more information is saved as:
# /home/jenkins/agent/workspace/line_8382-speed-up-product-build/com.sigasi.hdt.vhdl.test.projects/hs_err_pid27153.log
#
# Compiler replay data is saved as:
# /home/jenkins/agent/workspace/line_8382-speed-up-product-build/com.sigasi.hdt.vhdl.test.projects/replay_pid27153.log
#
# If you would like to submit a bug report, please visit:
#   https://github.com/adoptium/adoptium-support/issues
#
```

## Triaging info

Java version:

OpenJDK Runtime Environment Temurin-17.0.3+7 (17.0.3+7) (build 17.0.3+7)

What is your operating system and platform?

Amazon Linux release 2 (Karoo) on x86-64

How did you install Java?

Binary archive, tar.gz.

Did it work before?

We've been having this issue for months, since it relies on timings and code layouts there are periods in which we have many failures, then weeks without any.

Did you test with other Java versions?

Been having this since < Java 17. We haven't tried other VMs such as Graal or OpenJ9.

We've been faithfully upgrading to the latest Temurin version since Java 11, up to 12, 13, 14, 15 and now 17. (we skipped 16).

See our [hs_err_pid](https://drive.google.com/file/d/1B7HE5Wx4i8YifqEL81pAYdE5k-UjJ7iY/view?usp=sharing), [replay_pid](https://drive.google.com/file/d/1nt3ZwT_EBT1ol3zaFPAavoFRLflDZcXk/view?usp=sharing) and [core dump (2.3 GB)](https://drive.google.com/file/d/1T-y5TO3VUYH6o4p4laHoO3RnDlybmNZq/view?usp=sharing)

We're determined to help you help us. If there's anything more we can do, please let us know. We're trying to minimize this to a reproducible example, but that will take time, and definitely won't be easy due to the extreme flakiness of the failure.

**Comment by tivervac** [comment]
We don't use Lucene nor JavaFX, but that is the only link I found to `PhaseIdealLoop::build_loop_late_post_work` as well

**Comment by tivervac** [comment]
In the meantime, the linked bug has been split in two. [The first](https://bugs.openjdk.org/browse/JDK-8275610) appears irrelevant to us, we still encounter the issue with JustJ 17.0.4. [The second](https://bugs.openjdk.org/browse/JDK-8285835) is probably the one we need fixing.

**Comment by karianna** [comment]
17.0.4.1 should have https://bugs.openjdk.org/browse/JDK-8275610 fixed (the first)

**Comment by tellison** [comment]
Pinged about this at EclipseCon by @tivervac

**Comment by karianna** [comment]
Maintainers are commenting on the upstream issue, but it's still open for now.

**Comment by tivervac** [comment]
@uschindler That's quite possible. By now we've been able to (temporarily?) work around this issue by not obfuscating one of our classes. We obfuscate using [ZKM](https://www.zelix.com/klassmaster/index.html). That obfuscator likes to split off code and wrap it in other classes.

**Comment by uschindler** [comment]
So you don't have the source code of:
C2:  25179 18398       4       com.sigasi.hdt.vhdl.effectanalysis.d::visitIdentifierPathElement (97 bytes)

??? Too bad. Maybe a disassembly of bytecode?

**Comment by tivervac** [comment]
Sadly, I had this info at the time of writing, but not anymore at this point.

I'll see whether I can reproduce it again

**Comment by vans239** [comment]
Am I correct that upstream bug should be resolved in 19.0.2+7 ?
I feel that we faced same problem using 19.0.2+7

**Comment by karianna** [comment]
No fix version is for Java 20

**Comment by vans239** [comment]
@karianna How can I check it?
I thought that https://bugs.openjdk.org/browse/JDK-8297510 corresponds to issue above and should be resolved in 19.0.2 build 7

**Comment by karianna** [comment]
https://bugs.openjdk.org/browse/JDK-8285835

**Comment by vans239** [comment]
The link shows that the issue was backported
<img width="694" alt="image" src="https://user-images.githubusercontent.com/403189/215623323-9245941a-8ad7-4214-8d2c-8124ebfa5979.png">

**Comment by karianna** [comment]
Fair point, should be fixed then.  If you have a new crash log I can post it.

**Comment by karianna** [comment]
@vans239 Are you able to try the latest LTS or Java 20.0.1 and let us know if this is resolved for you?

**Comment by tivervac** [comment]
Sadly even with the backported bug fix mentioned above, we're still encountering the issue with `Temurin 17.0.7+7`

**Comment by vans239** [comment]
> Are you able to try the latest LTS or Java 20.0.1 and let us know if this is resolved for you?

We were still observing issues with 19.0.2+7. I am trying currently 20.0.1+9 and not able reproduce so far. Will have more info next week when we rollout it fully to prod

**Comment by karianna** [comment]
> Sadly even with the backported bug fix mentioned above, we're still encountering the issue with `Temurin 17.0.7+7`

I would try 17.0.8 JIC.

**Comment by karniemi** [comment]
Maybe related: https://github.com/openjdk/jdk/pull/15399 ... "SIGSEGV in PhaseIdealLoop::build_loop_late_post_work" ... but for a different reason. Backporting has not been even discussed on that one yet.

**Comment by tivervac** [comment]
Definitely possible, thanks for the link!

**Comment by lhotari** [comment]
Here are 2 crashes from Apache Pulsar, full hs_err_pid*.log files: https://gist.github.com/lhotari/53b72683ad4f339dfbcfd8b9b97062b9 .

```
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f927e8d5113, pid=3924, tid=4012
#
# JRE version: OpenJDK Runtime Environment Temurin-17.0.8.1+1 (17.0.8.1+1) (build 17.0.8.1+1)
# Java VM: OpenJDK 64-Bit Server VM Temurin-17.0.8.1+1 (17.0.8.1+1, mixed mode, sharing, tiered, compressed class ptrs, z gc, linux-amd64)
# Problematic frame:
# V  [libjvm.so+0xad5113]  PhaseIdealLoop::build_loop_late_post_work(Node*, bool)+0xe3
#
```
Happens in 17.0.8.1 . The Apache Pulsar issue is https://github.com/apache/pulsar/issues/19307 . Any help is appreciated.

**Comment by lhotari** [comment]
Looks like the previously posted GH issue links to https://bugs.openjdk.org/browse/JDK-8314024 which will be backported to 17.0.10 .
